std/algorithm/comparison.d

```diff
@@ -1565,3 +1568,197 @@ unittest
+
+        while (!saved.empty) {
+            lastItem = saved.front;
+
+            if (r1.save.count(lastItem) != r2.save.count(lastItem))
+            {
+                return false;
+            }
+
+            saved.popFront();
+
+            // Skip over runs of the same item
+            while (!saved.empty && saved.front == lastItem)
+            {
+                saved.popFront();
+            }
+        }
+
+        return true;
+    }
+}
+
+bool isPermutation(alias less = "a < b", Range1, Range2)
+    (SortedRange!(Range1, less) r1, SortedRange!(Range2, less) r2)
+if (!isInfinite!Range1 &&
+    !isInfinite!Range2)
+{
+    return r1 == r2;
+}
+
+///
+@safe pure unittest
+{
+    assert(isPermutation([1, 2, 3], [3, 2, 1]));
+    assert(isPermutation([1.1, 2.3, 3.5], [2.3, 3.5, 1.1]));
+    assert(isPermutation("abc", "bca"));
+
+    assert(isPermutation([1.1, 2.3, 3.5], [2.3, 3.5, 1.1], AllocateGC.yes));
+    assert(!isPermutation([1, 2], [3, 4], AllocateGC.yes));
+
+    assert(!isPermutation([1, 2], [3, 4]));
+    assert(!isPermutation([1, 1, 2, 3], [1, 2, 2, 3]));
+    assert(!isPermutation([1, 1], [1, 1, 1]));
+}
+
+/// Examples with sorted ranges
+@safe pure unittest
+{
+    import std.algorithm.sorting : sort;
+
+    assert(isPermutation(
+            sort([1, 2, 3]),
+            sort([3, 2, 1])
+    ));
+    assert(isPermutation(
+            sort([1.1, 2.3, 3.5]),
+            sort([2.3, 3.5, 1.1])
+    ));
+
+    // Works with any user defined sorting
+    assert(!isPermutation(
+            sort!"a > b"([1, 2]),
+            sort!"a > b"([3, 4])
+    ));
+
+    assert(isPermutation(
+        [1, 2, 3].assumeSorted,
+        [1, 2, 3].assumeSorted
+    ));
+
+    // Because the sorting is different for the ranges this will not use
+    // the sorted range overload
+    assert(isPermutation(
+        [3, 2, 1].assumeSorted!"a > b",
+        [1, 2, 3].assumeSorted!"a < b"
+    ));
+}
+
+@safe pure unittest
+{
+    import std.internal.test.dummyrange;
+
+    ReferenceForwardRange!int r1 = new ReferenceForwardRange!int([1, 2, 3, 4]);
+    ReferenceForwardRange!int r2 = new ReferenceForwardRange!int([1, 2, 4, 3]);
+    assert(isPermutation(r1, r2));
+
+    ReferenceForwardRange!int r3 = new ReferenceForwardRange!int([1, 2, 3, 4]);
+    ReferenceForwardRange!int r4 = new ReferenceForwardRange!int([4, 2, 1, 3]);
+    assert(isPermutation(r3, r4, AllocateGC.yes));
+
+    DummyRange!(ReturnBy.Reference, Length.Yes, RangeType.Random) r5;
+    DummyRange!(ReturnBy.Reference, Length.Yes, RangeType.Random) r6;
+    assert(isPermutation(r5, r6));
+}
+
+// Test CTFE
+@safe pure unittest
+{
+    enum result1 = isPermutation([1, 2, 3], [3, 2, 1]);
+    static assert(result1);
+
+    enum result2 = isPermutation(
+        [1, 2, 3].assumeSorted,
+        [1, 2, 3].assumeSorted
+    );
+    static assert(result2);
+
+    //enum result3 = isPermutation([1.1, 2.3, 3.5], [2.3, 3.5, 1.1], AllocateGC.yes);
+    //static assert(result3);
```

Comment:
Allocating with GC may interact badly with CTFE, perhaps? I'm not 100% sure.

Debugging functions in CTFE is a pain, because you can't use a debugger and the compiler may not be in a state that corresponds with an actual runtime state, making it complicated to figure out what went wrong. You can try `pragma(msg)`, but that may or may not give you what you want.